modules/xfem/src/dirackernels/XFEMPressure.C

```diff
@@ -0,0 +1,94 @@
+
+template<>
+InputParameters validParams<XFEMPressure>()
+{
+  InputParameters params = validParams<DiracKernel>();
+  params.addRequiredParam<unsigned int>("component", "The component for the pressure");
+  params.addParam<Real>("factor", 1.0, "The magnitude to use in computing the pressure");
+  params.addParam<FunctionName>("function", "The function that describes the pressure");
+  params.addParam<bool>("use_displaced_mesh", false, "Whether to use geometric search on displaced mesh.");
+  return params;
+}
+
+XFEMPressure::XFEMPressure(const InputParameters & parameters) :
+  DiracKernel(parameters),
+  _component(getParam<unsigned int>("component")),
+  _factor(getParam<Real>("factor")),
+  _function(isParamValid("function") ? &getFunction("function") : NULL),
+  _displaced(getParam<bool>("use_displaced_mesh"))
+{
+  FEProblem * fe_problem = dynamic_cast<FEProblem *>(&_subproblem);
+
+  if (fe_problem == NULL)
+    mooseError("Problem casting _subproblem to FEProblem in XFEMPressure");
+
+  if (!_displaced)
+  {
+    GeometricSearchData & geom_search_data = fe_problem->geomSearchData();
+    element_pair_locators = &geom_search_data._element_pair_locators;
+  }
+  else
+  {
+    GeometricSearchData & displaced_geom_search_data = fe_problem->getDisplacedProblem()->geomSearchData();
+    element_pair_locators = &displaced_geom_search_data._element_pair_locators;
+  }
+}
+
+void
+XFEMPressure::addPoints()
+{
+  _elem_point_normal.clear();
+  _elem_point_JxW.clear();
+
+  for (std::map<unsigned int, MooseSharedPointer<ElementPairLocator> >::iterator it_epl = element_pair_locators->begin();
+       it_epl != element_pair_locators->end();
+       ++it_epl)
+  {
+    ElementPairLocator & elem_pair_loc = *it_epl->second;
+    // go over pair elements
+    const std::list<std::pair<const Elem*, const Elem*> > & elem_pairs = elem_pair_loc.getElemPairs();
+    for (std::list<std::pair<const Elem *, const Elem*> >::const_iterator it_ep = elem_pairs.begin();
+         it_ep != elem_pairs.end(); ++it_ep)
+    {
+      const Elem * elem = it_ep->first;
+      const Elem * pair_elem = it_ep->second;
+      const ElementPairInfo & info = elem_pair_loc.getElemPairInfo(elem);
+
+      for (unsigned int i = 0; i < info._constraint_q_point.size(); ++i)
+      {
+        _elem_point_normal[elem][info._constraint_q_point[i]] = -info._normal;
+        _elem_point_normal[pair_elem][info._constraint_q_point[i]] = info._normal;
+        _elem_point_JxW[elem][info._constraint_q_point[i]] = info._constraint_JxW[i];
+        _elem_point_JxW[pair_elem][info._constraint_q_point[i]] = info._constraint_JxW[i];
+        addPoint(elem, info._constraint_q_point[i]);
+        addPoint(pair_elem, info._constraint_q_point[i]);
+      }
+    }
+  }
+}
+
+Real
+XFEMPressure::computeQpResidual()
+{
+  Real factor = _factor;
+
+  if (_function)
+    factor *= _function->value(_t, _current_point);
+
+  Point normal = _elem_point_normal[_current_elem][_current_point];
```

Comment:
It's generally a bad idea to use floating point values as keys in a map because it fails if they change slightly.  They probably wouldn't in this case, but it still seems like a bad idea. I think we still have a qp index in this case that we might be able to use to look this stuff up.